srcts/src/time/throttle.ts

```diff
@@ -19,40 +19,62 @@ class Throttler<X extends AnyVoidFunction> implements InputRatePolicy<X> {
     this.args = null;
   }
 
+  // If not timed out, immediately call the function and set the timer;
+  // if timed out, just queue up the args for the call when the timer runs out.
+  // Later calls during the same timeout will overwrite earlier ones
   normalCall(...args: Parameters<X>): void {
-    this.args = args;
+    this.args = args; // This will be an empty list (not null)
+    // if called without arguments, and `[null]` if called with `null`
+
+    // Only invoke immediately if there isn't a timer running
     if (this.timerId === null) {
-      this.$invoke();
-      this.timerId = setTimeout(() => {
-        // IE8 doesn't reliably clear timeout, so this additional
-        // check is needed
-        if (this.timerId === null) return;
-        this.$clearTimer();
-        if (args.length > 0) this.normalCall(...args);
-      }, this.delayMs);
+      this.$invoke(); // This also sets the timeout now
     }
   }
+
+  // Reset the timer if active and call immediately
   immediateCall(...args: Parameters<X>): void {
     this.$clearTimer();
     this.args = args;
     this.$invoke();
   }
+
+  // Is there a call waiting to send?
   isPending(): boolean {
-    return this.timerId !== null;
+    return this.args !== null;
```

Comment:
Note that this changes the semantics of `isPending()`, compared to the old version, before this code was refactored:
https://github.com/rstudio/shiny/blob/v1.6.0/srcjs/input_rate.js#L90-L92

I think the new semantics are correct, but I'll need to check whether this will cause any problems in existing code.